### ISSUE TITLE
Is there a way to hide/suppress any console output when loading and inferencing via PyTorch Hub?

### ISSUE THREAD

**Issue by haimat** [ultralytics/yolov5#2862]
## ❔Question
I would like to use YOLOv5 in a custom project, so I need to be in control of what will be printed to the console. Currently I am loading the model and using it like described in your wiki:

    model = torch.hub.load("ultralytics/yolov5", "custom", path_or_model=args.model)
    model.conf = args.confidence
    model.iou = args.iou
    img = Image.open(args,image)
    results = model(img, size=args.size, augment=True)
    preds = results.pandas().xyxy[0]

However, these line result in some console output (like network layout, autoShape info, etc.)
Is there any way to hide all these standad console output?

**Comment by glenn-jocher** [comment]
@haimat yes we probably want to group these outputs into a verbose=True setting. If you'd like to update the PyTorch Hub model code for this and submit a PR that would be awesome! If not we can add this to our TODO queue, which is very long at the moment.

**Comment by glenn-jocher** [comment]
@haimat see https://github.com/ultralytics/yolov5/blob/master/hubconf.py for the right place to start.

https://github.com/ultralytics/yolov5/blob/1df8c6c963d31ce84895101a70e45e0afdcb0bc2/hubconf.py#L1-L7

**Comment by haimat** [comment]
> @haimat yes we probably want to group these outputs into a verbose=True setting. If you'd like to update the PyTorch Hub model code for this and submit a PR that would be awesome! If not we can add this to our TODO queue, which is very long at the moment.

@glenn-jocher Well, for now I always worked with TF and Keras, so this is my first project with PyTorch and its hub. This YOLOv5 project is pretty cool though and I am always willing to learn something new, so why not, I can give this change a try. But could you please be so kind and give me just a few hints where to start and how you would like to have this change? I am totally fine with Python in general and these things, I would just like to know what you would have in mind, so that my change can reflect that. Thanks in advance!

**Comment by glenn-jocher** [comment]
@haimat well, there's two ways of managing the output, either with `if verbose: logger.info()` statements or maybe by setting the logging level directly before the logging events. The second is preferred if possible, though I've never implemented this myself so I'm not sure if it makes sense.

The first is pretty simple, but will require a lot of code changes, plus passing the verbose value through several functions to all of the places that currently print/log output, which would be a hassle. The main place where the model architecture is printed as it's built is here:
https://github.com/ultralytics/yolov5/blob/1df8c6c963d31ce84895101a70e45e0afdcb0bc2/models/yolo.py#L201-L202

**Comment by haimat** [comment]
@glenn-jocher Ok thanks, I will have a look at thia. One more thing though: When I update the logging logic e.g. for the `detect.py` file and related modules, then this would not directly reflect the hub model, right? How can I then check whether my change would also work for the hub version of YOLOv5?

**Comment by glenn-jocher** [comment]
@haimat ah well yes once we merge a PR, new operations with the same Hub cache would be unaffected unless the Hub cache is updated with `force_reload=True`. You can think of force_reload as a git pull for pytorch hub.

But also detect.py is mostly a separate inference channel from pytorch hub.

**Comment by NanoCode012** [comment]
Hi, speaking of the first method that Glenn mentioned, maybe we can rework the logging logic a bit? It was left untouched for a while.

https://github.com/ultralytics/yolov5/blob/1df8c6c963d31ce84895101a70e45e0afdcb0bc2/utils/general.py#L33-L36

Maybe we can a `verbose` option here? It could also be used with the `train.py` and `test.py` later as well by passing into this function.

```python
def set_logging(rank=-1, verbose=True): 
     logging.basicConfig( 
         format="%(message)s", 
         level=logging.INFO if (verbose and rank in [-1, 0]) else logging.WARN) 
```

https://github.com/ultralytics/yolov5/blob/1df8c6c963d31ce84895101a70e45e0afdcb0bc2/detect.py#L28-L29

The only thing to note is that if `verbose` is false, it should disable the majority/all the output (as far as I know).

**Comment by glenn-jocher** [comment]
@NanoCode012 yeah I think this is a great idea. I wonder if there's a way to pair this with a global verbose variable or an environment variable somehow so any function could set verbose=False and the other functions/files would be faithful to the change.

**Comment by haimat** [comment]
I was thinking of utilizing Python's logging functionalities. How about passing the log level (like "error", "warning", "info") to the script, and based on that more or less logging will be printed. And all that comes for free if we replace all `print` statements with the logger call.

**Comment by glenn-jocher** [comment]
@haimat @NanoCode012 yes these are both good ideas. I've added a TODO label here to remind ourselves to get to this eventually, but I don't have much availability for this and the priority is not high, so it may take a while. If either of you want to take a stab at it feel free to submit a PR, thank you!

**Comment by glenn-jocher** [comment]
@haimat good news 😃! Your original issue may now been fixed ✅ in PR #2926 by @NanoCode012.  To receive this update you can:
- `git pull` from within your `yolov5/` directory
- `git clone https://github.com/ultralytics/yolov5` again
- Force-reload PyTorch Hub: `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by glenn-jocher** [comment]
Removing TODO as this is partly resolved now by avoiding model construction during most PyTorch Hub model loads.

**Comment by haimat** [comment]
Hi, I have switched to the `yolov5` module from pyPI via `pip install yolov5`, and with that the default output is gone.  
This is fine for me now, thanks for your help.

**Comment by JonathanSamelson** [comment]
Hi, for those who want no logs at all, you can use this piece of code:

```
import logging

logging.getLogger("utils.general").setLevel(logging.WARNING)  # yolov5 logger
```

This is probably not the most intuitive logger name though.
Hope it helps.

**Comment by glenn-jocher** [comment]
@JonathanSamelson thanks for the tip! Where would this code be placed? Before or after importing LOGGER from utils.general?

@kalenmike example of eliminating all YOLOv5 logs in https://github.com/ultralytics/yolov5/issues/2862#issuecomment-1016516835

This might be useful for creating a global verbose mode.

**Comment by JonathanSamelson** [comment]
I simply use it at the very beginning of my code. Since I wanted to infer the results without anything printed in the console.
I don't import LOGGER from the module `utils.general`, but doing so setup something in the `logging `package even before the logger is registered, which is quite nice to mute everything at once.

For instance:

```
import torch
import logging
import cv2

if __name__ == '__main__':
    logging.getLogger("utils.general").setLevel(logging.WARNING)
    model = torch.hub.load('ultralytics/yolov5', 'custom', path="path/to/yolov5s.pt")
    img = cv2.imread("path/to/image.png)[..., ::-1]
    results = model(img)
    results.print()
    results.show()
```

Usually it works with other loggers, but they have a more explicit name:
```
    logging.getLogger("matplotlib").setLevel(logging.WARNING)
    logging.getLogger("tensorflow").setLevel(logging.WARNING)
```

However I saw that this is setup with `__name__` here, which results in `utils.general`.

**Comment by glenn-jocher** [comment]
@JonathanSamelson ah ok got it. Is there anyway to rename the logger? i.e. to something like this? 
```python
    logging.getLogger("yolov5").setLevel(logging.WARNING)
```

BTW we just did a PR https://github.com/ultralytics/yolov5/pull/6353 that sets `verbose=False` for YOLOv5 with the VERBOSE environment variable:

```bash
export VERBOSE=False
python detect.py
```

<img width="1194" alt="Screen Shot 2022-01-20 at 12 26 22 PM" src="https://user-images.githubusercontent.com/26833433/150432134-1025a3b2-83d2-46cf-b4ba-5b138cc9cffc.png">

**Comment by JonathanSamelson** [comment]
@glenn-jocher  Yes, there is. 

Instead of :

```
def set_logging(name=None, verbose=VERBOSE):
    # Sets level and returns logger
    if is_kaggle():
        for h in logging.root.handlers:
            logging.root.removeHandler(h)  # remove all handlers associated with the root logger object
    rank = int(os.getenv('RANK', -1))  # rank in world for Multi-GPU trainings
    logging.basicConfig(format="%(message)s", level=logging.INFO if (verbose and rank in (-1, 0)) else logging.WARNING)
    return logging.getLogger(name)


LOGGER = set_logging(__name__)  # define globally (used in train.py, val.py, detect.py, etc.)
```

I'd simply use: 

```
LOGGER = set_logging("yolov5")
```

And I think that's it because internally you probably import LOGGER everytime you use it. Externally you would change the log level exactly has you mentionned: 

`logging.getLogger("yolov5").setLevel(logging.WARNING)`

I don't mind opening/amending a PR but as I'm very new to this project I don't know if there is any other impacts or what has been done exactly with the VERBOSE environment variable.

**Comment by glenn-jocher** [comment]
@JonathanSamelson oh perfect! That seems like an easy change. Can you check that this doesn't alter the current logging output and then please submit a PR with this fix?

**Comment by JonathanSamelson** [comment]
Here we go @glenn-jocher : #6421 
As I wrote, I checked that it doesn't alter the current output of the logger by using this line above in detect.py. It indeed keeps the same information if the logger is set to INFO and hides everything with WARNING (as there is no warning 😉).
Feel free to do more tests, don't hesitate if you need something else.

**Comment by glenn-jocher** [comment]
@JonathanSamelson thank you, PR is merged!
